File: frontend/src/services/costsRaw.ts
/**
 * @file src/services/costsRaw.ts
 * @description Warstwa serwisowa dla modułu "Koszty ILUO (beta)" (tabela costs_raw / ERP).
 *              Styl spójny z services/costs.ts.
 *
 * KROK 2b: oddzial (KOD legacy) + oddzial_display (nazwa) z backendu.
 * KROK 3:  odpowiedź listy niesie stan flag produkcyjnych.
 * KROK 4:  status przypisania w nagłówku + metoda assign().
 */

// Płaski nagłówek dokumentu (lista)
export interface CostRawHeader {
  id: number;
  numer: string | null;
  data: string | null;
  nip: string | null;
  nazwa_skrocona: string | null;
  netto: number | null;
  vat: number | null;
  brutto: number | null;
  etykieta: string | null;
  punkt_handlowy: string | null;
  oddzial: string | null;          // KOD legacy (np. "Pcim", "HQ", "MG", "Private")
  oddzial_display: string | null;  // nazwa wyświetlana (np. "Centrala", "Sklep internetowy")
  numer_obcy: string | null;
  liczba_pozycji: number;
  prowizja: boolean;  // dokument prowizyjny (Prowizja - KOSZT) — właściciel tylko Oddział/PH
  // Status przypisania (krok 4)
  assigned_cost_id: number | null; // cost_id w all_costs; null = nieprzypisany
  assigned_at: string | null;
  assigned_by: string | null;
}

// Odpowiedź listy (paginowana)
export interface CostRawListResponse {
  data: CostRawHeader[];
  total: number;
  total_sum: number;
  total_sum_netto: number;
  limit: number;
  offset: number;
  // Flagi produkcyjne (źródło: backend) — patrz routes/costs_raw.py
  require_label: boolean;
  date_filter_enabled: boolean;
  min_date: string | null;
}

// Pojedyncza pozycja dokumentu (modal)
export interface CostRawPosition {
  indeks: string | null;
  nazwa: string | null;
  ilosc: number | null;
  cena: number | null;
  vat: number | null;
}

// Szczegóły dokumentu: nagłówek + pozycje
export interface CostRawDetail {
  header: CostRawHeader;
  pozycje: CostRawPosition[];
}

// Body przypisania własności (krok 4)
export interface CostRawAssignRequest {
  cost_own: string;          // Wspólny | Oddział | Centrala | Przedstawiciel
  cost_ph: string | null;    // wymagany dla "Przedstawiciel"
  author: string;
}

// Wynik przypisania (krok 4)
export interface CostRawAssignResponse {
  cost_id: number;
  assigned_at: string;
  cost_branch: string;
  cost_mo: number;
  cost_year: number;
  cost_value: number;
}

class CostsRawService {
  private readonly apiUrl = '/api/costs-iluo';

  // Lista dokumentów (same nagłówki, paginowane)
  async getList(params: {
    oddzial?: string;
    wyklucz_oddzialy?: string;
    szukaj?: string;
    nazwa_like?: string;
    numer_like?: string;
    ma_etykiete?: boolean;
    przypisane?: boolean;
    data_od?: string;
    data_do?: string;
    sort_by?: string;
    sort_dir?: 'asc' | 'desc';
    limit?: number;
    offset?: number;
  }): Promise<CostRawListResponse> {
    try {
      const queryParams = new URLSearchParams();

      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null && value !== '') {
          queryParams.append(key, value.toString());
        }
      });

      const response = await fetch(`${this.apiUrl}?${queryParams.toString()}`);

      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.detail || `Błąd podczas pobierania kosztów ILUO: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error('Wystąpił nieznany błąd podczas pobierania kosztów ILUO');
    }
  }

  // Szczegóły jednego dokumentu (nagłówek + pozycje) — dla modala
  async getDetail(costId: number): Promise<CostRawDetail> {
    try {
      const response = await fetch(`${this.apiUrl}/${costId}`);

      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.detail || `Błąd podczas pobierania szczegółów dokumentu: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error('Wystąpił nieznany błąd podczas pobierania szczegółów dokumentu');
    }
  }

  // Przypisanie własności kosztu -> tworzy rekord w all_costs (krok 4)
  async assign(costId: number, body: CostRawAssignRequest): Promise<CostRawAssignResponse> {
    try {
      const response = await fetch(`${this.apiUrl}/${costId}/assign`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      });

      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.detail || `Błąd podczas przypisywania kosztu: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error('Wystąpił nieznany błąd podczas przypisywania kosztu');
    }
  }

  // Lista unikalnych KODÓW oddziałów (legacy cost_branch) — do dropdownu filtra
  async getBranches(): Promise<string[]> {
    try {
      const response = await fetch(`${this.apiUrl}/branches`);

      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.detail || `Błąd podczas pobierania oddziałów: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error('Wystąpił nieznany błąd podczas pobierania oddziałów');
    }
  }
}

export const costsRawService = new CostsRawService();

File: backend/src/models/costs_raw.py
# models/costs_raw.py
"""
Model dla tabeli costs_raw — surowe koszty z ERP (ILUO).
Źródło prawdy dla modułu "Koszty ILUO (beta)".

Struktura odpowiada DDL:
    CREATE TABLE costs_raw (
        id                BIGSERIAL    PRIMARY KEY,
        naglowek          JSONB        NOT NULL,
        pozycje           JSONB        NOT NULL,
        created_at        TIMESTAMPTZ  NOT NULL DEFAULT now(),
        updated_at        TIMESTAMPTZ  NOT NULL DEFAULT now(),
        -- KROK 4a (2026-07): status przypisania do all_costs
        assigned_cost_id  BIGINT       NULL,
        assigned_at       TIMESTAMPTZ  NULL,
        assigned_by       VARCHAR(100) NULL
    );

UWAGA: kolumny JSON-owe są surowe (raw) — nie rozbijamy ich na pola.
- naglowek: numer, data, nip, nazwa_skrocona, brutto, netto, vat,
            etykieta, punkt_handlowy, numer_obcy
- pozycje:  tablica obiektów { indeks, nazwa, ilosc, cena, vat }

assigned_cost_id = cost_id rekordu w all_costs po przypisaniu własności;
NULL = dokument nieprzypisany (w puli "do przypisania").
"""

from sqlalchemy import Column, BigInteger, DateTime, String, func
from sqlalchemy.dialects.postgresql import JSONB

from database import Base


class CostsRaw(Base):
    __tablename__ = "costs_raw"

    id = Column(BigInteger, primary_key=True, index=True)
    naglowek = Column(JSONB, nullable=False)
    pozycje = Column(JSONB, nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    # KROK 4a: status przypisania do all_costs
    assigned_cost_id = Column(BigInteger, nullable=True, index=True)
    assigned_at = Column(DateTime(timezone=True), nullable=True)
    assigned_by = Column(String(100), nullable=True)

    def __repr__(self):
        numer = None
        if isinstance(self.naglowek, dict):
            numer = self.naglowek.get("numer")
        return f"<CostsRaw(id={self.id}, numer='{numer}', assigned={self.assigned_cost_id})>"